stl/inc/ranges

```diff
@@ -872,6 +872,133 @@ namespace ranges {
+        constexpr explicit common_view(_Rng&& _Range) noexcept(
+            is_nothrow_constructible_v<_Vw, views::all_t<_Rng>&&>) // strengthened
+            : _Base(views::all(_STD forward<_Rng&&>(_Range))) {}
+        // clang-format on
+
+        _NODISCARD constexpr _Vw base() const& noexcept(
+            is_nothrow_copy_constructible_v<_Vw>) /* strengthened */ requires copy_constructible<_Vw> {
+            return _Base;
+        }
+        _NODISCARD constexpr _Vw base() && noexcept(is_nothrow_move_constructible_v<_Vw>) /* strengthened */ {
+            return _STD move(_Base);
+        }
+
+        _NODISCARD constexpr auto begin() noexcept(
+            _RANGES begin(_Base) && is_nothrow_move_constructible_v<iterator_t<_Vw>>) {
+            if constexpr (random_access_range<_Vw> && sized_range<_Vw>) {
+                return _RANGES begin(_Base);
+            } else {
+                return common_iterator<iterator_t<_Vw>, sentinel_t<_Vw>>{_RANGES begin(_Base)};
+            }
+        }
+
+        _NODISCARD constexpr auto begin() const
+            noexcept(_RANGES begin(_Base) && is_nothrow_move_constructible_v<iterator_t<const _Vw>>) {
+            if constexpr (random_access_range<const _Vw> && sized_range<const _Vw>) {
+                return _RANGES begin(_Base);
+            } else {
+                return common_iterator<iterator_t<const _Vw>, sentinel_t<const _Vw>>{_RANGES begin(_Base)};
+            }
+        }
+
+        _NODISCARD constexpr auto end() noexcept(
+            _RANGES end(_Base) && is_nothrow_move_constructible_v<iterator_t<_Vw>>) {
+            if constexpr (random_access_range<_Vw> && sized_range<_Vw>) {
+                return _RANGES begin(_Base) + _RANGES size(_Base);
+            } else {
+                return common_iterator<iterator_t<_Vw>, sentinel_t<_Vw>>{_RANGES end(_Base)};
+            }
+        }
+
+        _NODISCARD constexpr auto end() const
+            noexcept(_RANGES end(_Base) && is_nothrow_move_constructible_v<iterator_t<const _Vw>>) {
+            if constexpr (random_access_range<const _Vw> && sized_range<const _Vw>) {
+                return _RANGES begin(_Base) + _RANGES size(_Base);
+            } else {
+                return common_iterator<iterator_t<const _Vw>, sentinel_t<const _Vw>>{_RANGES end(_Base)};
+            }
+        }
+
+        _NODISCARD constexpr auto size() noexcept(
+            noexcept(_RANGES size(_Base))) /* strengthened */ requires sized_range<_Vw> {
+            return _RANGES size(_Base);
+        }
+        _NODISCARD constexpr auto size() const
+            noexcept(noexcept(_RANGES size(_Base))) /* strengthened */ requires sized_range<const _Vw> {
+            return _RANGES size(_Base);
+        }
+    };
+
+    template <class _Rng>
+    common_view(_Rng &&) -> common_view<views::all_t<_Rng>>;
+
+    namespace views {
+        // VARIABLE views::common
+        template <class _Rng>
+        concept _Can_view_all = requires(_Rng&& __r) {

```

Comment:
This shouldn't be unnecessary: `viewable_range<T>` is meant to admit exactly those types `T` for which `views::all(declval<T>())` is well-formed.

Reply:
Double negation?

Reply:
Just to be sure I follow you. The standard says [this](http://eel.is/c++draft/range.common#overview-3.1)
>views​::​all(E), if decltype((E)) models common_­range and views​::​all(E) is a well-formed expression.

So you are saying that when `decltype((E))` models `common_range` I can assume that `views::all` is well formed?

Reply:
> Double negation?

Ouch - sorry - edited in place.

> Just to be sure I follow you. The standard says [this](http://eel.is/c++draft/range.common#overview-3.1)
> 
> > views​::​all(E), if decltype((E)) models common_­range and views​::​all(E) is a well-formed expression.
> 
> So you are saying that when `decltype((E))` models `common_range` I can assume that `views::all` is well formed?

The poorly-expressed intent of the Standard is that range adaptors and range adaptor closures constrain their first argument with `viewable_range` before doing anything else:
* [range.adaptor.object]/1: "A _range adaptor closure object_ is a unary function object that accepts a `viewable_range` argument and returns a `view`."
* [range.adaptor.object]/2: "A _range adaptor object_ is a customization point object that accepts a `viewable_­range` as its first argument and returns a `view`."

The wording for `views::common` apparently didn't get the memo about the blanket requirement so it's applying it again in a slightly different form.

In any case, `views::common` has a `viewable_range` constraint on line 982 which guarantees that `views::all(declval<_Rng>())` is well-formed, so the application of `_Can_view_all` on line 968 is redundant.